### ISSUE TITLE
Support Windows (cmd/Powershell)

### ISSUE THREAD

**Issue by vko-online** [denisidoro/navi#40]
Please add support to Windows (Choco, curl)

**Comment by denisidoro** [comment]
Hi! I'm not very familiar with the Windows ecosystem. What setup do you have exactly? WSL? Cygwin? Git for windows?

In any of these cases, installing via git + make should work out of the box and the problem is limited to add a package to choco, right? 

Otherwise, I don't plan to support the vanilla command prompt or Powershell, because navi relies heavily on tools that only exist on Unix-like systems.

The only hope I see for this case is a tool out there able to compile bash scripts to a Windows binary, but that probably doesn't exist.

**Comment by denisidoro** [comment]
navi is now written in Rust. It should be relatively trivial to start supporting Windows.

AFAIK, the only change needed is to use cmd/Powershell instead of bash to generate variable suggestions.

**Comment by denisidoro** [comment]
Hi! Sorry for the late update.

Being 100% honest here: this is a feature it would be nice to have but unfortunately I haven't had much free time lately, so I have no plans to implement this in the near future.

If anyone could submit a PR for this, I'll gladly accept it, though! 👍

**Comment by denisidoro** [comment]
Work on this front is mostly, if not all, done. In theory, navi should be buildable for the `x86_64-pc-windows-gnu` target now.

**Comment by denisidoro** [comment]
`bash` still is needed under the hood, but it should be possible to use it under Cygwin now. 

(In WSL as well, but that has always been the case)

**Comment by Milo123459** [comment]
Can't wait! Maybe also publish to [scoop](https://scoop.sh)?

**Comment by denisidoro** [comment]
Scoop: https://github.com/ScoopInstaller/Main/pull/2100

Choco: https://github.com/denisidoro/navi/issues/532

Pre-compiled binary: https://github.com/denisidoro/navi/releases

I'd love to hear feedback from Windows users, specially because I don't use Windows so I can't test navi on it.

**Comment by denisidoro** [comment]
BTW, bash is still necessary under the hood but, if there's enough demand for it, navi could work entirely under cmd.exe. Please let me know if there's a use case for that.

**Comment by Milo123459** [comment]
I mean I think a lot of people want it on windows because of the "command fixing" system, which is literally amazing. 

Thanks for the other info.

Oh and, it can work completely under cmd or powershell.

**Comment by kai2nenobu** [comment]
For chocolatey users, [navi 2.15.1 package](https://community.chocolatey.org/packages/navi/2.15.1) has been approved.  Try it out and if something fails during installation, feel free to [report issue in the package repository](https://github.com/kai2nenobu/chocolatey-packages).

[Package for version 2.16.0](https://community.chocolatey.org/packages/navi/2.16.0) is under moderation and may be approved in few days.

**Comment by Milo123459** [comment]
The PR for scoop is out of date, so I'm making one for this scoop bucket: https://github.com/kidonng/sushi

Btw, making it run entirely on cmd.exe would be best.

**Comment by denisidoro** [comment]
You can already try running navi entirely on cmd.exe by setting "cmd.exe" as `shell.command` in the config file: https://github.com/denisidoro/navi/blob/master/docs/config_file_example.yaml

Some stuff will be broken (such as git clone and copy to clipboard) but I can fix reported issues until navi is completely agnostic to a POSIX shell.

TL;DR: beta testers wanted

**Comment by Milo123459** [comment]
Added Navi to my custom scoop repo `Milo123459/cone`.

**Comment by Milo123459** [comment]
If you get the time, please make a widget for powershell.

**Comment by denisidoro** [comment]
I have zero experience with Powershell :(

I created https://github.com/denisidoro/navi/issues/570 with the label "help wanted", though.

If someone submits a PR, I'll gladly review/merge it!

**Comment by kboom** [comment]
PowerShell 7+ is absolutely amazing. Would be indeed great to have it. Easy to learn too.

**Comment by kboom** [comment]
In fact it works! You just have to change https://github.com/denisidoro/navi/blob/master/docs/config_file_example.yaml to use powershell, and use choco to install it. Should be better documented, as this is great!

**Comment by denisidoro** [comment]
Sorry for the confusion.

PowerShell should be working for a while, indeed.

What I meant is that it navi currently doesn't support a PoweShell *shell widget* (see https://github.com/denisidoro/navi/issues/40)

**Comment by kboom** [comment]
It doesn't work well with variables in the case of PowerShell. When I select a command that expects a variable:

![image](https://user-images.githubusercontent.com/3975111/171654389-721d8964-3345-4a33-a0ec-fc3d99ca6b78.png)

**Comment by alexis-opolka** [comment]
@HUMORCE Sorry to revive this old thread but is your issue still persisting with the latest version of navi?
You can currently specify the config path for navi with the `NAVI_CONFIG` environment variable.

For more details, see: https://github.com/denisidoro/navi/blob/master/docs/navi_config.md#config-file-path

If this does not resolve your issue of specifying the config path, please open a dedicated github issue which should help sort features request.

**Comment by HUMORCE** [comment]
@alexis-opolka I'll respond and try solving it once I'm free. I'm thinking maybe next week.

**Comment by alexis-opolka** [comment]
Any news on your side @HUMORCE ?

**Comment by HUMORCE** [comment]
> You can currently specify the config path for navi with the NAVI_CONFIG environment variable.

It's work.

**Comment by alexis-opolka** [comment]
> > You can currently specify the config path for navi with the NAVI_CONFIG environment variable.
> 
> It's work.

It's nice to hear that! :smile:

Can we consider this issue solved then?

**Comment by HUMORCE** [comment]
In fact, you should submit the application to choco or Powershell gallery. But an env var can solve many problems for package managers, you should ask for the OP instead me.

I will not make any changes of the app in Scoop's manifest, unless more users need persistence of the config file.

**Comment by alexis-opolka** [comment]
There is already a package for Choco (see #532 and https://community.chocolatey.org/packages/navi) although it is not an official distribution, I don't think navi has really a place in the PowerShell gallery though.

If it works for you and you don't need to change anything that's also a pretty good outcome. Don't hesitate to ping me one day if there is an issue with navi and Scoop, even more if it's on Windows as it seems I'm the only one of the currently active contributors able to work on Windows. (At least to my knowledge)

@vko-online Was your original issue solved? Would you need anything else?
If your issue is solved, I would appreciate it if you could close it, thanks. :smile:

**Comment by alexis-opolka** [comment]
I'm closing this issue, don't hesitate to re-open it if needed.
